tsl/src/remote/dist_copy.c

```diff
@@ -306,54 +321,122 @@ send_end_binary_copy_data(PGconn *connection)
+				 errmsg("%s", primary),
+				 detail == NULL ? 0 : errdetail("%s", detail),
+				 hint == NULL ? 0 : errhint("%s", hint)));
+}
+
 static void
 finish_outstanding_copies(CopyConnectionState *state)
 {
 	ListCell *lc;
-	List *results = NIL;
+	List *volatile results = NIL;
+	PGresult *volatile res = NULL;
 
-	foreach (lc, state->connections_in_use)
+	PG_TRY();
 	{
-		TSConnection *conn = lfirst(lc);
-		PGconn *pg_conn = remote_connection_get_pg_conn(conn);
-		PGresult PG_USED_FOR_ASSERTS_ONLY *res;
+		foreach (lc, state->connections_in_use)
+		{
+			TSConnection *conn = lfirst(lc);
+			PGconn *pg_conn = remote_connection_get_pg_conn(conn);
 
-		if (state->using_binary)
-			if (send_end_binary_copy_data(pg_conn) != 1)
+			if (state->using_binary)
+				if (send_end_binary_copy_data(pg_conn) != 1)
+					ereport(ERROR,
+							(errcode(ERRCODE_CONNECTION_EXCEPTION),
+							 errmsg("%s", PQerrorMessage(pg_conn))));
+
+			if (PQputCopyEnd(pg_conn, NULL) == -1)
 				ereport(ERROR,
 						(errcode(ERRCODE_CONNECTION_EXCEPTION),
 						 errmsg("%s", PQerrorMessage(pg_conn))));
 
-		if (PQputCopyEnd(pg_conn, NULL) == -1)
-			ereport(ERROR,
-					(errcode(ERRCODE_CONNECTION_EXCEPTION), errmsg("%s", PQerrorMessage(pg_conn))));
+			res = PQgetResult(pg_conn);
+			Assert(res != NULL);
+			results = lappend(results, res);
+
+			/* Need to get result a second time to move the connection out of copy mode */
+			res = PQgetResult(pg_conn);
 
-		results = lappend(results, PQgetResult(pg_conn));
-		/* Need to get result a second time to move the connection out of copy mode */
-		res = PQgetResult(pg_conn);
+			if (res != NULL)
+				ereport(ERROR,
+						(errcode(ERRCODE_INTERNAL_ERROR),
+						 errmsg("COPY command resulted in unexpected state")));
+		}
+	}
+	PG_CATCH();
+	{
 		if (res != NULL)
-			ereport(ERROR,
-					(errcode(ERRCODE_INTERNAL_ERROR),
-					 errmsg("COPY command resulted in unexpected state")));
+			PQclear(res);
+
+		clear_results(results, false);
+		PG_RE_THROW();
 	}
+	PG_END_TRY();
 
-	foreach (lc, results)
-		if (PQresultStatus(lfirst(lc)) != PGRES_COMMAND_OK)
-			ereport(ERROR,
-					(errcode(ERRCODE_INTERNAL_ERROR),
-					 errmsg("error during copy completion: %s", PQresultErrorMessage(lfirst(lc)))));
+	Assert(res == NULL);
+	clear_results(results, true);
 }
 
 static List *
-get_connections_for_chunk(CopyConnectionState *state, Chunk *chunk)
+get_connections_for_chunk(RemoteCopyContext *context, const Chunk *chunk)
 {
+	CopyConnectionState *state = &context->connection_state;
+	MemoryContext oldmctx;
 	ListCell *lc;
+	List *conns;
 
 	foreach (lc, state->cached_connections)
-		if (((ChunkConnectionList *) lfirst(lc))->chunk == chunk)
-			return ((ChunkConnectionList *) lfirst(lc))->connections;
+	{
+		ChunkConnectionList *chunkconns = lfirst(lc);
 
-	return create_connection_list_for_chunk(state, chunk)->connections;
+		if (chunkconns->chunk_id == chunk->fd.id)
+			return chunkconns->connections;
```

Comment:
This is definitely hit in my benchmark (I added print statement here). But I guess our regression tests never get to the point of actually switching between cached chunks. I'll try to see what it would take to make that happen.